Sources/SwiftParser/Parser.swift

```diff
@@ -424,6 +424,58 @@ extension Parser {
       self.missingToken(.identifier, text: nil)
     )
   }
+
+  /// Expect a right brace but with a little smart recovery logic:
+  /// If `leftBrace` is missing, we only recover to a `}` whose indentation is greater or equal to that of `introducer`.
+  /// That way, if the developer forgot to to type `{`, we won't eat `}` that were most likely intended to close an outer scope.
+  ///
+  /// If `leftBrace` is present or `introducer` is `nil`, this is equivalent to `self.expect(.rightBrace)`.
+  mutating func expectRightBrace(leftBrace: RawTokenSyntax, introducer: RawTokenSyntax?) -> (unexpected: RawUnexpectedNodesSyntax?, token: RawTokenSyntax) {
+    func indentation(_ pieces: [RawTriviaPiece]) -> RawTriviaPiece? {
+      if pieces.last?.isNewline == true {
+        return .spaces(0)
+      }
+      guard pieces.count >= 2 else {
+        return nil
+      }
+      guard pieces[pieces.count - 2].isNewline else {
+        return nil
+      }
+      switch pieces[pieces.count - 1] {
+      case .spaces, .tabs:
+        return pieces[pieces.count - 1]
+      default:
+        return nil
+      }
+    }
+
+    guard leftBrace.isMissing, let introducer = introducer else {
+      // Fast path for correct parses: If leftBrace is not missing, just `expect`.
+      return self.expect(.rightBrace)
+    }
+
+    var lookahead = self.lookahead()
+    guard let recoveryHandle = lookahead.canRecoverTo([.rightBrace]) else {
+      // We can't recover to '}'. Synthesize it.
+      return (nil, self.missingToken(.rightBrace, text: nil))
+    }
+
+    // We can recover to a '}'. Decide whether we want to eat it based on its indentation.
```

Comment:
Yes, I know. I thought about hard-coding a tab-width of 4 or something like that but in the end I decided that if you used mixed tabs and spaces, I don’t care about you.

Reply:
Ouch, no love for the mixed-tab-and-spaces crowd I see 😅. Since this just makes the recovery slightly nicer, I'm fine with that. Someone can always come by and add a tab width later.